File: 2022/Day16.cpp
#include <bits/stdc++.h>
using namespace std;

using num_t = unsigned long;
using memo_key_t = tuple<num_t, num_t>;

struct hash_tuple {
    template <class T1, class T2>
    size_t operator()(const tuple<T1, T2>& x) const {
      return get<0>(x) ^ get<1>(x);
    }
};

using Memo = unordered_map<memo_key_t, short, hash_tuple>;

constexpr num_t N = 26 * 52 * 52 * (1l << 15);

num_t walk(num_t n, Memo& memo, vector<num_t>& flows, vector<vector<num_t>>& neighbours, num_t k, num_t p1, num_t p2, num_t o) {
  if (k == 0) return 0;
  memo_key_t key(k << 20l | p1 << 10l | p2, o);
  if (auto x = memo.find(key); x != memo.end()) return x->second;

  num_t res = 0;

  if (k > 1) {
    vector<tuple<num_t, num_t>> choices1, choices2;

    // if current valve can be opened
    if ((not (o & (1l << p1))) && flows[p1] > 0)
        choices1.emplace_back(make_tuple(p1, o | (1l << p1)));
    for (auto& n : neighbours[p1])
        choices1.emplace_back(make_tuple(n, o));
    if ((not (o & (1l << p2))) && flows[p2] > 0)
        choices2.emplace_back(make_tuple(p2, o | (1l << p2)));
    for (auto& n : neighbours[p2])
        choices2.emplace_back(make_tuple(n, o));

    for (auto& [np1, no1] : choices1) {
        for (auto& [np2, no2] : choices2) {
          num_t x = walk(n, memo, flows, neighbours, k-1, np1, np2, no1 | no2);
          if (x > res) res = x;
      }
    }
  }

  for (num_t i = 0; i < n; ++i) {
      res += flows[i] * !!(o & (1l << i));
  }
  memo[key] = res;

  if (memo.size() % 1000000 == 0)
      cout << (memo.size() / (float) N * 100) << "% | " << memo.size() << " / " << N << endl;

  return res;
}

num_t solve(num_t n, vector<num_t> flows, vector<vector<num_t>> neighbours, num_t start) {
  Memo memo;

  // walk(graph, k, position, opened_valves) returns the maximum flow achievable in k minutes
  return walk(n, memo, flows, neighbours, 26, start, start, 0);
}

int main() {
  num_t n_ex = 10;
  vector<num_t> flows_ex = { 0, 13, 2, 20, 3, 0, 0, 22, 0, 21 };
  vector<vector<num_t>> neighbours_ex = {{3, 8, 1}, {2, 0}, {3, 1}, {2, 0, 4}, {5, 3}, {4, 6}, {5, 7}, {6}, {0, 9}, {8}};
  num_t start_ex = 0;

  num_t n = 52;
  vector<num_t> flows = { 0, 21, 0, 0, 16, 0, 0, 0, 10, 0, 17, 0, 0, 24, 0, 0, 0, 0, 20, 0, 0, 13, 0, 0, 0, 0, 0, 5, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 25, 0, 0, 0, 0, 0, 18, 0, 0, 9, 0, 23, 12, 3 };
  vector<vector<num_t>> neighbours = {{14, 36}, {46, 39}, {51, 14}, {34, 50}, {29}, {15, 35}, {25, 50}, {27, 44}, {20, 45, 19}, {39, 21}, {28, 25}, {41, 47}, {31, 51}, {42, 26}, {2, 0, 37, 17, 40}, {5, 47}, {51, 29}, {14, 48}, {46}, {27, 8}, {21, 8}, {20, 23, 9, 36}, {42, 44}, {21, 27}, {51, 43}, {6, 10}, {13, 49}, {19, 23, 40, 7}, {33, 10}, {16, 4}, {47, 38}, {12, 35}, {51, 50}, {28, 49}, {35, 3}, {34, 5, 31, 37}, {0, 21}, {14, 35}, {30}, {9, 1}, {27, 14}, {50, 11}, {22, 13}, {24, 47}, {22, 45, 7}, {44, 8}, {1, 18}, {48, 30, 43, 11, 15}, {17, 47}, {33, 26}, {3, 41, 6, 32}, {2, 16, 24, 12, 32}};
  num_t start = 14;

  printf("Solution for part 2: %lu\n", solve(n_ex, flows_ex, neighbours_ex, start_ex));
  printf("Solution for part 2: %lu\n", solve(n, flows, neighbours, start));
  return 0;
}

// Runs in 810 secs lol

// Solution part 1: 1862
// Solution part 2: 2422

File: 2022/Day16.py
#! /usr/bin/env python
# pylint: disable=unused-wildcard-import
from util import *

def parse(inp):
    # return structuredre(inp, r"Valve ([A-Z]{2}) has flow rate=(\d+); tunnels lead to valves (.+)", (str, int, str))
    xs = structuredre(inp, r"Valve ([A-Z]{2}) has flow rate=(\d+); tunnel[a-z ]+([A-Z,\s]+)+", (str, int, str))
    for i, l in enumerate(lines(inp)):
        xs[i] = list(xs[i])
        if len(l.split("valves ")) == 1:
            xs[i][2] = [l.split("valve ")[1]]
        else:
            xs[i][2] = l.split("valves ")[1].split(", ")

    return xs

def p1(inp):
    # xs = list(sorted(inp, key=lambda x: x[2]))
    nodes = [(x[0], [(y, x[1]) for y in x[2]]) for x in inp]
    # real_nodes = []
    # for i in range(30):
        # for n in nodes:
    names = [x[0] for x in inp]
    # nodes = flatten([(x[0] + ' x', x[1]), (x[0] + ' o', x[1])] for x in nodes)
    print(nodes)
    print(names)
    g = nx_from_node_list(nodes, True, True)
    flow = dict()
    for name, f, _ in inp:
        flow[name] = f
    print(g['AA'])
    # nx_draw_graph(g)

    memo = dict()
    # walk(graph, k, position, opened_valves) returns the maximum flow achievable in k minutes
    def walk(g, k, p, o):
        nonlocal memo
        h = ''.join(sorted(o))
        if (k, p, h) in memo:
            return memo[(k, p, h)]

        # print(k, p)
        if k == 0: return 0
        choices = []

        # if current valve can be opened
        if p not in o and flow[p] > 0:
            choices.append(walk(g, k-1, p, o.union(set([p]))))
        # else:
        for n in g[p]:
            # print(n)
            choices.append(walk(g, k-1, n, o))

        f = 0
        for x in o:
            f += flow[x]
        res = max(c + f for c in choices)
        memo[(k, p, h)] = res
        return res
        # print(g, k, p, o)

    res = walk(g, 30, 'AA', set())

    return res

def p2(inp):
    names = [x[0] for x in inp]
    flows = [x[1] for x in inp]
    g = [[names.index(y) for y in x[2]] for x in inp]
    n_valves = len(inp)
    start = names.index('AA')

    print(flows)
    print(g)
    print(n_valves)
    print(start)

    N = 26 * 52**2 * 2**15
    memo = dict()
    # walk(graph, k, position, opened_valves) returns the maximum flow achievable in k minutes
    def walk(g, k, p1, p2, o):
        nonlocal memo
        if (k, p1, p2, o) in memo:
            return memo[(k, p1, p2, o)]
        if k == 0: return 0
        choices1, choices2 = [], []

        # if current valve can be opened
        if (not (o & (1 << p1))) and flows[p1] > 0:
            choices1.append((p1, o | (1 << p1)))
        for n in g[p1]:
            choices1.append((n, o))
        if (not (o & (1 << p2))) and flows[p2] > 0:
            choices2.append((p2, o | (1 << p2)))
        for n in g[p2]:
            choices2.append((n, o))

        res = 0
        for np1, no1 in choices1:
            for np2, no2 in choices2:
                x = walk(g, k-1, np1, np2, no1 | no2)
                if x > res:
                    res = x

        for i in range(n_valves):
            res += flows[i] * (1 if o & (1 << i) else 0)
        memo[(k, p1, p2, o)] = res

        if len(memo) % 1000000 == 0:
            print(f"{len(memo) / N * 100}% | {len(memo)} / {N}")


        return res

    res = walk(g, 26, start, start, 0)

    return res

ex = """Valve AA has flow rate=0; tunnels lead to valves DD, II, BB
Valve BB has flow rate=13; tunnels lead to valves CC, AA
Valve CC has flow rate=2; tunnels lead to valves DD, BB
Valve DD has flow rate=20; tunnels lead to valves CC, AA, EE
Valve EE has flow rate=3; tunnels lead to valves FF, DD
Valve FF has flow rate=0; tunnels lead to valves EE, GG
Valve GG has flow rate=0; tunnels lead to valves FF, HH
Valve HH has flow rate=22; tunnel leads to valve GG
Valve II has flow rate=0; tunnels lead to valves AA, JJ
Valve JJ has flow rate=21; tunnel leads to valve II
"""

def main():
    inp = parse(data())
    # inp = parse(ex)
    print(f"Solution for part 1:\n{p1(inp)}")
    # print(f"Solution for part 2:\n{p2(inp)}")

if __name__ == "__main__":
    main()

# Solution part 1: 1862
# Solution part 2: 2422
